### ISSUE TITLE
Header/footer temporary file is removed before wkhtmltopdf is called

### ISSUE THREAD

**Issue by hpoo** [mileszs/wicked_pdf#1031]
## Issue description
When creating a pdf-file with content given in footer- and header-options, the pdf creation sometimes fail. 
The reason for the failure is that temporary files widget-pdf creates from the header/footer content are removed before the `wkhtmltopdf` binary is called. `wkhtmltopdf` fails with errors like
`Error: Failed to load file:////tmp/wicked_header_pdf20220620-3625428-872dqb.html?`

Further investigating the issue it seems that the culprit may be that only the temporary file name is stored to options hash in the `options_parser#parse_header_footer`. The temporary file objects are not stored and are loose for the JRuby garbage collection to clean up and remove. 

## Expected or desired behavior
It would be expected that the temporary file live long enough to be available for the `wkhtmltopdf` to use

## System specifications
wicked_pdf gem version:wicked_pdf-2.6.3
wkhtmltopdf version:wkhtmltopdf 0.12.4
platform/distribution and version (e.g. Windows 10 / Ubuntu 16.04 / Heroku cedar):
ubuntu 18.04, Rails 6.0 with JRuby 9.3.3.0

**Comment by thanhcuong1990** [comment]
I got the same issue with:
`wicked_pdf: 2.6.3`
`wkhtmltopdf 0.12.6 (with patched qt)`
`macOS Montery`

The header/ footer html files is deleted before the pdf generated. 
`wicked_header_pdfxxxxx.html`
`wicked_footer_pdfxxxxxx.html`

**Comment by GalenkoEugene** [comment]
I got the same issue with:
`ruby:2.6.6-slim`
`wicked_pdf:  2.1.0 and 2.6.3`

**Comment by unixmonkey** [comment]
The tempfiles are pushed into an [instance variable](https://github.com/mileszs/wicked_pdf/blob/master/lib/wicked_pdf/pdf_helper.rb#L110) that only gets [cleared after the render action](https://github.com/mileszs/wicked_pdf/blob/master/lib/wicked_pdf/pdf_helper.rb#L9). 

How do you think that is happening early?

**Comment by thanhcuong1990** [comment]
@unixmonkey I was render header/footer by `render_to_string` described in [Super Advanced Usage](https://github.com/mileszs/wicked_pdf#super-advanced-usage). Like below:

```ruby
# create a pdf from string using templates, layouts and content option for header or footer
pdf = WickedPdf.new.pdf_from_string(
  render_to_string('templates/pdf', layout: 'pdfs/layout_pdf.html'),
  footer: {
    content: render_to_string(
  		'templates/footer',
  		layout: 'pdfs/layout_pdf.html'
  	)
  }
)
```

The header/footer temp files are pushed in [option_parser.rb](https://github.com/mileszs/wicked_pdf/blob/master/lib/wicked_pdf/option_parser.rb#L64).

I was try to debug then see the header/footer temp files will deleted at https://github.com/mileszs/wicked_pdf/blob/master/lib/wicked_pdf.rb#L64

**Comment by hpoo** [comment]
Also my case uses `pdf_from_string` and the temporary files are created in the `OptionParser`, not the `PdfHelper`. The `OptionParser` object lives only for the one[ line](https://github.com/mileszs/wicked_pdf/blob/3492a51b840913333b06c8f0aa8130bff6cd6ba4/lib/wicked_pdf.rb#L102). The instance and its' instance variables are free food for the JRuby garbage collection once the line is done. 
Both `PdfHelper` and `OptionsParser` do similar option parsing and use the same instance variable name, which may contribute to the confusion.

**Comment by tomascco** [comment]
I've got the same issue with `WickedPdf.new.pdf_from_string`, using WickedPdf 2.6.3 and Ruby 2.6.9.

**Comment by Holist** [comment]
Same here...
```ruby:2.7.2```
```wicked_pdf: 2.6.3```
```wkhtmltopdf-binary: 0.12.5```
Any workaround ?

**Comment by vtllr** [comment]
I have the same issue: our PDFs on production (Ubuntu 20.04.2 LTS) are randomly generating an error: sometimes, the PDF is loaded correctly, sometimes not.

The error is the same: "Failed to load file:////tmp/wicked_footer_pdf20220825 [...]" so I suspect the cache is destroyed before the PDF is generated properly. Can't replicate on my development env tho (Mac)

```
ruby:2.7.2
wicked_pdf: 2.6.3
wkhtmltopdf-binary: 0.12.6.5
```

**Comment by thanhcuong1990** [comment]
@hpoo, @GalenkoEugene, @tomascco, @vtllr, @Holist I was try to create a PR #1039 to resolved this issue. It's should work with my project now. 
Can you can try to play with my PR to see if it works?

**Comment by andrebnf** [comment]
@thanhcuong1990 thanks for the fix proposal! I confirm that your changes fix the issues we had in production environments (issue I'm referring to is the same reported here: https://github.com/mileszs/wicked_pdf/issues/1031#issuecomment-1227807796)

**Comment by waqarnazir** [comment]
same issue please merge and release

**Comment by waqarnazir** [comment]
Same issue, and also being reproduced in development environment as well (PDF generation is being done in background). Kindly merge and release the fix.

**Comment by stevesizer** [comment]
any update on this?

**Comment by thanhcuong1990** [comment]
@stevesizer This issue is resolved in PR https://github.com/mileszs/wicked_pdf/pull/1039. 

You can do a temporary fix in your `Gemfile`.
`gem 'wicked_pdf', git: 'https://github.com/thanhcuong1990/wicked_pdf.git', branch: 'master'`

**Comment by ehutzelman** [comment]
I ran into this issue as well and it was more apparent under ruby 3.2 - I guess the GC is more aggressive which resulted in these files getting cleaned up faster after `OptionParser` goes out of scope. I think we can still use temp files here, but we need to ensure the reference to `OptionParser` (which contains the temp files for header/footer) doesn't go out of scope before `WickedPdf` has finished its work.

I've been running the following fix for the past couple days and it seems to fix my issues with the temp files being purged too quickly:

Original:
https://github.com/mileszs/wicked_pdf/blob/f7663737ab3f1171f1a4ff8179d924c47c4c32d6/lib/wicked_pdf.rb#L101-L103

Updated:
```ruby
  def parse_options(options)
    # Store reference to option parser in instance var to ensure internal temp files remain
    # referenced for life of WickedPdf instance
    @option_parser = OptionParser.new(binary_version)
    @option_parser.parse(options)
  end
```

Fork: https://github.com/ehutzelman/wicked_pdf

**Comment by camilova** [comment]
Hi, thanks for the patch!, it solves the problem, but in my case I was facing two more problems:

- On `Windows` it raises `Permission Denied` for created temp files because they were not closed after creation.
- If one reuses the `WickedPdf` class it fails because old tempfiles paths persist in the array `hf_tempfiles`, failing when later tries deleting that files that don't exists anymore.

I create a PR #1054 that solves both cases and could help someone.
